autoload/lsp/utils/text_edit.vim

```diff
@@ -2,74 +2,66 @@ function! lsp#utils#text_edit#apply_text_edits(uri, text_edits) abort
-  if a:text_edit.range.end.line <= a:cursor_position.line
-      let l:cursor_offset = l:new_lines_len - (a:text_edit.range.end.line - a:text_edit.range.start.line) - 1
-      let a:cursor_position.line += l:cursor_offset
-  endif
+    " create before/after line.
+    let l:start_line = getline(a:text_edit['range']['start']['line'] + 1)
+    let l:end_line = getline(a:text_edit['range']['end']['line'] + 1)
+    let l:before_line = strcharpart(l:start_line, 0, a:text_edit['range']['start']['character'])
+    let l:after_line = strcharpart(l:end_line, a:text_edit['range']['end']['character'], strchars(l:end_line) - a:text_edit['range']['end']['character'])
+
+    " create new lines.
+    let l:new_lines = lsp#utils#_split_by_eol(a:text_edit['newText'])
+    let l:new_lines[0] = l:before_line . l:new_lines[0]
+    let l:new_lines[-1] = l:new_lines[-1] . l:after_line
+
+    " fixendofline
+    let l:buffer_length = len(getbufline(a:bufnr, '^', '$'))
+    let l:should_fixendofline = lsp#utils#buffer#_get_fixendofline(a:bufnr)
+    let l:should_fixendofline = l:should_fixendofline && l:new_lines[-1] ==# ''
+    let l:should_fixendofline = l:should_fixendofline && l:buffer_length <= a:text_edit['range']['end']['line']
+    let l:should_fixendofline = l:should_fixendofline && a:text_edit['range']['end']['character'] == 0
+    if l:should_fixendofline
+        call remove(l:new_lines, -1)
+    endif
 
-  " append new lines.
-  call append(a:text_edit.range.start.line, l:new_lines)
+    let l:new_lines_len = len(l:new_lines)
+    let l:range_len = (a:text_edit['range']['end']['line'] - a:text_edit['range']['start']['line']) + 1
+
+    " fix cursor pos
+    if a:text_edit['range']['end']['line'] <= a:cursor_position['line'] && a:text_edit['range']['end']['character'] <= a:cursor_position['character']
+        if a:text_edit['range']['end']['line'] == a:cursor_position['line']
+            let l:end_character = strchars(l:new_lines[-1]) - strchars(l:after_line)
+            let l:end_offset = a:cursor_position['character'] - a:text_edit['range']['end']['character']
+            let a:cursor_position['character'] = l:end_character + l:end_offset
+        endif
+        let a:cursor_position['line'] += l:new_lines_len - l:range_len
+    endif
 
-  " remove old lines
-  execute printf('%s,%sdelete _',
-  \   l:new_lines_len + a:text_edit.range.start.line + 1,
-  \   min([l:new_lines_len + a:text_edit.range.end.line + 1, line('$')])
-  \ )
+    " append or delete lines.
+    if l:new_lines_len > l:range_len
+        call append(a:text_edit['range']['start']['line'], repeat([''], l:new_lines_len - l:range_len))
+    elseif l:new_lines_len < l:range_len
+        let l:offset = l:range_len - l:new_lines_len
+        call s:delete(a:bufnr, a:text_edit['range']['start']['line'] + 1, a:text_edit['range']['start']['line'] + l:offset)
+    endif
 
-  return l:cursor_offset
+    " set lines.
+    call setline(a:text_edit['range']['start']['line'] + 1, l:new_lines)
```

Comment:
This changes aims to keep the `virtual_text`/`marker` etc.

The previous approach was simple but always append lines so shifted line-oriented information unexpectedly.